### PR TITLE
fix: don't publish flow-client TS sources in JAR

### PR DESCRIPTION
**Pull request by Haprog**
Improve performance of webpack build by only shipping precompiled JS of the flow-frontend TS sources, so webpack doesn't try to use the TS sources which is slightly slower than using JS (and `.d.ts`) directly. The TS sources shouldn't be needed in the JAR anyway and this has the benefit of making the JAR a bit smaller in size also. Related to #10281.

### CODE REVIEW

**Comment by Legioth** [comment]
I guess one drawback of this is that it would give worse source maps for anyone trying to locally debug anything related to the code for which the original TS files are missing.

**Comment by Haprog** [comment]
> I guess one drawback of this is that it would give worse source maps for anyone trying to locally debug anything related to the code for which the original TS files are missing.

I don't think that's an issue since we have:
```json
    "sourceMap": true,
    "inlineSources": true,
```
in `tsconfig.json` (for the flow-client build)

The `.js.map` files seem to contain the full TS sources too.

File: flow-client/src/main/frontend/Flow.ts [hunk 1]
```diff
@@ -251,7 +251,7 @@ export class Flow {
       const { appId } = appConfig;
 
       // Load bootstrap script with server side parameters
-      const bootstrapMod = await import('./FlowBootstrap');
+      const bootstrapMod = await import('../resources/META-INF/resources/frontend/FlowBootstrap');
```
Comment (haijian-vaadin) [inline flow-client/src/main/frontend/Flow.ts:254]: Tried the snapshot build, got frontend compile errors 
ERROR in ../target/flow-frontend/Flow.js
Module not found: Error: Can't resolve '../resources/META-INF/resources/frontend/FlowBootstrap'

**Review by haijian-vaadin — CHANGES_REQUESTED** [review]

**Comment by Haprog** [comment]
After moving also `FlowClient` and `FlowBootstrap` (`.js` and `.d.ts`) it seems to work now. I tested locally in a Flow app. Let's see if the CI also passes now.

**Comment by Haprog** [comment]
I forgot to mvn clean before installing flow-client locally when testing so I didn't catch some problems. New try now. Tested locally working again.

File: flow-client/.eslintrc.js [hunk 2]
```diff
@@ -18,6 +18,13 @@ module.exports = {
       },
     },
 
+    {
+      "files": ["**/.eslintrc.js"],
+      "globals": {
+        "module": "readonly",
+      },
+    },
```
Comment (Haprog) [inline flow-client/.eslintrc.js:26]: Add an entry for `.eslintrc.js` files so it doesn't complain about `module.exports = {` line in the eslint configs.

File: flow-client/scripts/.eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  "env": {
+    "browser": false,
+    "node": true,
```
Comment (Haprog) [inline flow-client/scripts/.eslintrc.js:4]: New eslint config for the `scripts` dir with `node` env so it doesn't complain about usage of `require()` in node scripts.

File: flow-client/scripts/client.js [hunk 4]
```diff
@@ -35,10 +36,13 @@ ${clientSource}
...
-fs.writeFileSync(toSourceDir + toFile, clientSource, 'utf8');
+fs.writeFileSync(sourceDir + toFile, clientSource, 'utf8');
 
-// Write to target
-fs.writeFileSync(toTargetDir + toFile, clientSource, 'utf8');
+// Write to target (copy '.d.ts' and '.js' files from sourceDir)
+mkdirp(targetDir);
+fs.readdirSync(sourceDir)
+  .filter(s => s.endsWith('.d.ts') ||s.endsWith('.js'))
+  .forEach(file => fs.copyFileSync(sourceDir + file, targetDir + file));
```
Comment (Haprog) [inline flow-client/scripts/client.js:45]: Use `mkdirp` to create this target dir as it's not automatically populated by Maven anymore due to moving the source files.

Now we also need to copy the FlowClient and FlowBootstrap files (`.js` and `.d.ts`) from source to target dir due to moving the source files. `tsc` (TS compiler) outputs the compiled results of TS files into target but doesn't copy the JS sources for these files that aren't written in TS. Here we just copy all `.d.ts` and `.js` from source to target to be more future proof.

Previously this script was only writing the generated `FlowClient.js` into both source and target. Now it's written into source first and then it gets copied to target also as part of this step.

File: flow-client/src/main/frontend/Flow.ts (hunk 1 above)
Comment (Haprog) [reply flow-client/src/main/frontend/Flow.ts]: Problem fixed.

**Comment by haijian-vaadin** [comment]
I just tried, now it indeed works. Good job!

**Review by haijian-vaadin — APPROVED** [review]

**Pull request merged**